### PR TITLE
Force hadoop client to hadoopcli on python3

### PR DESCRIPTION
**Pull request by gpoulin**
Also disable snakebite test for python3

### CODE REVIEW

File: luigi/mock.py [hunk 1]
```diff
@@ -170,6 +171,22 @@ def skip(func):
...
+def skipIf(condition):
+    """
+
+    """
+    if condition:
+        return skip
+    return lambda x: x
+
+
+skipIfPy3k = skipIf(six.PY3)
```
Comment (erikbern) [inline luigi/mock.py]: Why did you put it in mock.py?

Comment (gpoulin) [inline luigi/mock.py]: `skip` decorator was already there. It makes to keep those 2 together, no? Maybe `skip`shouldn't be there at the first place. You prefer that I move those `skip`, `skipIf` and `skipIfPy3k` somewhere in /test

Comment (erikbern) [inline luigi/mock.py]: Yeah I think that makes sense. But if you are busy we can merge now and fix it later.

**Comment by gpoulin** [comment]
@erikbern I moved the `skip` to test/helpers.py. It's not used a lot so it was better the fix now.

**Comment by erikbern** [comment]
Awesome! Any idea why the Hadoop tests are failing? I wonder if we need to bundle some module with the hadoop jobs.

**Comment by gpoulin** [comment]
The issue was really strange. The only fact of importing helpers was making the test fail. I think it was some issue with pickling and subprocess (helpers was not importable), but I'm not sure. However adding a simple __init__.py to the test folder seems to solve the issue.

**Comment by gpoulin** [comment]
Adding an __init__.py to the test folder cause other test to fail. I'm giving up, I put back the skip family in `luigi.mock` adding a `_` to those decorator. I prefer to spend time fixing test rather than skipping them.

File: luigi/hdfs.py [hunk 2]
```diff
@@ -616,6 +616,13 @@ def get_configured_hdfs_client(show_warnings=True):
     """
     config = hdfs()
     custom = config.client
+    if six.PY3 and (custom == "snakebite" or config.use_snakebite):
+        if show_warnings:
+            warnings.warn(
+                "snakebite client not compatible with python3 at the moment"
+                "falling back on hadoopcli"
```
Comment (berkerpeksag) [inline luigi/hdfs.py]: needs `stacklevel=2`

File: test/_s3_test.py [hunk 3]
```diff
@@ -39,20 +42,18 @@
...
-    mock_s3 = skip
+# moto break stuff under python3.4
+skip_tests = skip_tests or sys.version_info[:2] == (3, 4)
 
 
 AWS_ACCESS_KEY = "XXXXXXXXXXXXXXXXXXXX"
 AWS_SECRET_KEY = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
 
 
+@_skipIf(skip_tests)
```
Comment (berkerpeksag) [inline test/_s3_test.py]: I don't understand this function. Can't you use a `unittest.skipIf` decorator instead?

```py
@unittest.skipIf(sys.version_info[:2] == (3, 4))
class TestS3Target(unittest.TestCase):
    # ...
```

File: test/_s3_test.py [hunk 4]
```diff
@@ -39,20 +42,18 @@
     # https://github.com/spulec/moto/issues/29
     print('Skipping %s because moto does not install properly before '
           'python2.7' % __file__)
-    from luigi.mock import skip
-    mock_s3 = skip
+    skip_tests = True
```
Comment (berkerpeksag) [inline test/_s3_test.py]: Looks like https://github.com/spulec/moto/issues/29 has been already fixed.

File: test/_s3_test.py (hunk 3 above)
Comment (gpoulin) [inline test/_s3_test.py]: `unittest.skipIf` is not documented for python 2.6. I assume it's not compatible.

**Comment by erikbern** [comment]
That's fine, you can leave them in luigi.mock. I can try to factor it out

File: test/_s3_test.py (hunk 3 above)
Comment (berkerpeksag) [inline test/_s3_test.py]: Ah, I see. Thanks for the clarificaiton. You can import unittest2 conditionally. unittest2 will be installed if you're on Python 2.6: https://github.com/spotify/luigi/blob/master/tox.ini#L9

Comment (gpoulin) [inline test/_s3_test.py]: Thanks for pointing out. I'll modify appropriately.

**Comment by erikbern** [comment]
This is great except the merge conflict now :(

**Comment by gpoulin** [comment]
@erikbern I fixed the conflict. I also incorporated @berkerpeksag's comments (no more `luigi.mock.skip`). Also, #763 was skipping all hadoop test under python3. With this PR we only skip the test that actually require snakebite.

**Comment by gpoulin** [comment]
@berkerpeksag unittest2 doesn't seem to be available for python 2.6. The hadoop_test are failing with 
```pytb
@unittest.skipIf(six.PY3, "snakebite minicluster is not supported in python3 yet.")
AttributeError: 'module' object has no attribute 'skipIf'
```

even thought I put the following at the beginning of the file:

```python
try:
    import unittest2 as unittest
except ImportError:
    import unittest
```

**Comment by gpoulin** [comment]
I'm not familiar with tox, but the my understanding is that in the following configuration `py26` doesn't refer to the python version but to the tox environment. Since we are using travis to manage python version, there isn't any tox environment matching py26.

```ini
[testenv]
deps=
  py26: unittest2
```

To fix this issue, I added unittest2 to requirement and use unitest2 only when the corresponding function doesn't exist in unittest

```python
try:
    from unittest import skipIf
except ImportError:
    from unittest2 import skipIf
```

**Comment by gpoulin** [comment]
@erikbern With this last commit all tests that doesn't require Minicluster or snakebite are working on python3. I think we can close #382.

`test_multiple_workers` fails when it's run with all other tests, but it passes when run individually.

**Comment by Tarrasch** [comment]
Please also remove the "allowed failures" as they are not allowed anymore. https://travis-ci.org/spotify/luigi/builds/51701754

**Comment by Tarrasch** [comment]
Please also remove the "allowed failures" as they are not allowed anymore. https://travis-ci.org/spotify/luigi/builds/51701754

Very good job on this @gpoulin !! I guess I could merge this now. Right?

**Comment by Tarrasch** [comment]
Hmm, why is there only one test running? Is there only one non-snakebite&non-minicluster test for our hadoop stuff?

https://travis-ci.org/spotify/luigi/jobs/51701764

**Comment by gpoulin** [comment]
@Tarrasch There is indeed only one hdfs test that doesn't require Minicluster. The other "hadoop" tests are local test and are run as part of the nonhdfs environment. I'll try to see if I can make minicluster run under python3.

It's good to be merged.

**Comment by erikbern** [comment]
Awesome

**Pull request merged**